Shadow Manager Exception when writing to DB

**Describe the bug**
I have had the latest ShadowManager and Nucleus deployed on a device I have been developing on for a few days. Today, all my components that relied on reading shadows failed due to the ShadowManager component entering the broken state. I was able to get a stack trace from Greengrass's logs that show that the ShadowManager was broken because of an exception thrown [here](https://github.com/aws-greengrass/aws-greengrass-shadow-manager/blob/63a00b4d6b83bbae3a05c18ace5213a4a44dcb35/src/main/java/com/aws/greengrass/shadowmanager/ShadowManagerDAOImpl.java#L373). See the logs pasted below for more details. Because shadows are a core part of our system this render much of the device's functionality broken and is considered a critical failure.

**To Reproduce**
I was not able to reproduce this issue although it persisted until I downgraded my Nucleus back to 2.9.5 and Shadow Manager to 2.2.4. After deploying those versions the shadows became synced between the cloud and device and read/write-able as expected. 

Rebooting the device, power cycling the device, `systemctl restart`-ing `greengrass`, reinstalling the ShadowManager component via the LocalDebugConsole, re-deploying the same version of ShadowManager, and redeploying the components without ShadowManager then re-adding it in a second deployment all yielded the same results (logs shown below).

__Note:__ This broken state was first incurred during the 6/13 us-east-1 outage. My device is on us-east-1. 

**Environment**
 - OS: NVIDIA Jetpack 4.3
 - JDK version: 11.8.0
 - Nucleus version: 2.10.2
 - Shadow Manager version: 2.3.2

Greengrass logs captured after hitting "Reinstall" on the local debug console. The exception is thrown 3 times and then the component enters BROKEN. 

```
2023-06-13T20:35:28.417Z [INFO] (WebSocketWorker-62) com.aws.greengrass.localdebugconsole.SimpleHttpServer: Requested reinstall from aws.greengrass.ShadowManager. {serviceName=aws.greengrass.LocalDebugConsole, currentState=RUNNING}
2023-06-13T20:35:28.418Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.shadowmanager.ShadowManager: service-set-state. {serviceName=aws.greengrass.ShadowManager, currentState=BROKEN, newState=NEW}
2023-06-13T20:35:28.460Z [INFO] (pool-2-thread-23) org.flywaydb.core.internal.database.DatabaseFactory: Database: jdbc:h2:/greengrass/v2/work/aws.greengrass.ShadowManager/shadow (H2 1.4). {}
2023-06-13T20:35:28.522Z [INFO] (pool-2-thread-23) org.flywaydb.core.internal.command.DbValidate: Successfully validated 1 migration (execution time 00:00.049s). {}
2023-06-13T20:35:28.563Z [INFO] (pool-2-thread-23) org.flywaydb.core.internal.command.DbMigrate: Current version of schema "PUBLIC": 1. {}
2023-06-13T20:35:28.572Z [INFO] (pool-2-thread-23) org.flywaydb.core.internal.command.DbMigrate: Schema "PUBLIC" is up to date. No migration necessary.. {}
2023-06-13T20:35:28.645Z [INFO] (pool-2-thread-23) com.aws.greengrass.shadowmanager.ShadowManager: config. {syncStrategy=realTime, serviceName=aws.greengrass.ShadowManager, currentState=NEW}
2023-06-13T20:35:28.645Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.shadowmanager.ShadowManager: service-set-state. {serviceName=aws.greengrass.ShadowManager, currentState=NEW, newState=INSTALLED}
2023-06-13T20:35:28.659Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.shadowmanager.ShadowManager: service-set-state. {serviceName=aws.greengrass.ShadowManager, currentState=INSTALLED, newState=STARTING}
2023-06-13T20:35:28.663Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.shadowmanager.ShadowManager: service-set-state. {serviceName=aws.greengrass.ShadowManager, currentState=STARTING, newState=RUNNING}
2023-06-13T20:35:28.700Z [ERROR] (pool-2-thread-24) com.aws.greengrass.shadowmanager.ShadowManager: service-errored. {serviceName=aws.greengrass.ShadowManager, currentState=RUNNING}
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.FutureTask@65eae692[Not completed, task = com.aws.greengrass.shadowmanager.ShadowManagerDAOImpl$$Lambda$469/0x000000010044c840@484c05b7] rejected from java.util.concurrent.ThreadPoolExecutor@112efa7a[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 1]
	at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2055)
	at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:825)
	at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1355)
	at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:140)
	at com.aws.greengrass.shadowmanager.ShadowManagerDAOImpl.executeWriteOperation(ShadowManagerDAOImpl.java:373)
	at com.aws.greengrass.shadowmanager.ShadowManagerDAOImpl.insertSyncInfoIfNotExists(ShadowManagerDAOImpl.java:343)
	at com.aws.greengrass.shadowmanager.ShadowManager.insertSyncInfoIfNotPresent(ShadowManager.java:437)
	at com.aws.greengrass.shadowmanager.ShadowManager.initializeSyncInfo(ShadowManager.java:432)
	at com.aws.greengrass.shadowmanager.ShadowManager.startSyncingShadows(ShadowManager.java:537)
	at com.aws.greengrass.shadowmanager.ShadowManager.setupSync(ShadowManager.java:396)
	at com.aws.greengrass.shadowmanager.ShadowManager.startup(ShadowManager.java:480)
	at com.aws.greengrass.lifecyclemanager.Lifecycle.lambda$handleStateTransitionStartingToRunningAsync$9(Lifecycle.java:591)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)

2023-06-13T20:35:28.702Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.shadowmanager.ShadowManager: service-set-state. {serviceName=aws.greengrass.ShadowManager, currentState=RUNNING, newState=ERRORED}
2023-06-13T20:35:28.735Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.status.FleetStatusService: fss-status-update-published. Status update published to FSS. {trigger=ERRORED_COMPONENT, serviceName=FleetStatusService, currentState=RUNNING}
2023-06-13T20:35:28.738Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.shadowmanager.ShadowManager: service-set-state. {serviceName=aws.greengrass.ShadowManager, currentState=ERRORED, newState=STOPPING}
2023-06-13T20:35:28.799Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.shadowmanager.ShadowManager: service-set-state. {serviceName=aws.greengrass.ShadowManager, currentState=STOPPING, newState=INSTALLED}
2023-06-13T20:35:28.803Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.shadowmanager.ShadowManager: service-set-state. {serviceName=aws.greengrass.ShadowManager, currentState=INSTALLED, newState=STARTING}
2023-06-13T20:35:28.816Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.shadowmanager.ShadowManager: service-set-state. {serviceName=aws.greengrass.ShadowManager, currentState=STARTING, newState=RUNNING}
2023-06-13T20:35:28.831Z [ERROR] (pool-2-thread-24) com.aws.greengrass.shadowmanager.ShadowManager: service-errored. {serviceName=aws.greengrass.ShadowManager, currentState=RUNNING}
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.FutureTask@68ea093a[Not completed, task = com.aws.greengrass.shadowmanager.ShadowManagerDAOImpl$$Lambda$469/0x000000010044c840@60033dd5] rejected from java.util.concurrent.ThreadPoolExecutor@112efa7a[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 1]
	at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2055)
	at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:825)
	at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1355)
	at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:140)
	at com.aws.greengrass.shadowmanager.ShadowManagerDAOImpl.executeWriteOperation(ShadowManagerDAOImpl.java:373)
	at com.aws.greengrass.shadowmanager.ShadowManagerDAOImpl.insertSyncInfoIfNotExists(ShadowManagerDAOImpl.java:343)
	at com.aws.greengrass.shadowmanager.ShadowManager.insertSyncInfoIfNotPresent(ShadowManager.java:437)
	at com.aws.greengrass.shadowmanager.ShadowManager.initializeSyncInfo(ShadowManager.java:432)
	at com.aws.greengrass.shadowmanager.ShadowManager.startSyncingShadows(ShadowManager.java:537)
	at com.aws.greengrass.shadowmanager.ShadowManager.setupSync(ShadowManager.java:396)
	at com.aws.greengrass.shadowmanager.ShadowManager.startup(ShadowManager.java:480)
	at com.aws.greengrass.lifecyclemanager.Lifecycle.lambda$handleStateTransitionStartingToRunningAsync$9(Lifecycle.java:591)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)

2023-06-13T20:35:28.832Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.shadowmanager.ShadowManager: service-set-state. {serviceName=aws.greengrass.ShadowManager, currentState=RUNNING, newState=ERRORED}
2023-06-13T20:35:28.834Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.status.FleetStatusService: fss-status-update-published. Status update published to FSS. {trigger=ERRORED_COMPONENT, serviceName=FleetStatusService, currentState=RUNNING}
2023-06-13T20:35:28.839Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.shadowmanager.ShadowManager: service-set-state. {serviceName=aws.greengrass.ShadowManager, currentState=ERRORED, newState=STOPPING}
2023-06-13T20:35:28.864Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.shadowmanager.ShadowManager: service-set-state. {serviceName=aws.greengrass.ShadowManager, currentState=STOPPING, newState=INSTALLED}
2023-06-13T20:35:28.867Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.shadowmanager.ShadowManager: service-set-state. {serviceName=aws.greengrass.ShadowManager, currentState=INSTALLED, newState=STARTING}
2023-06-13T20:35:28.871Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.shadowmanager.ShadowManager: service-set-state. {serviceName=aws.greengrass.ShadowManager, currentState=STARTING, newState=RUNNING}
2023-06-13T20:35:28.901Z [ERROR] (pool-2-thread-24) com.aws.greengrass.shadowmanager.ShadowManager: service-errored. {serviceName=aws.greengrass.ShadowManager, currentState=RUNNING}
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.FutureTask@19487583[Not completed, task = com.aws.greengrass.shadowmanager.ShadowManagerDAOImpl$$Lambda$469/0x000000010044c840@4a4bc7fd] rejected from java.util.concurrent.ThreadPoolExecutor@112efa7a[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 1]
	at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2055)
	at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:825)
	at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1355)
	at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:140)
	at com.aws.greengrass.shadowmanager.ShadowManagerDAOImpl.executeWriteOperation(ShadowManagerDAOImpl.java:373)
	at com.aws.greengrass.shadowmanager.ShadowManagerDAOImpl.insertSyncInfoIfNotExists(ShadowManagerDAOImpl.java:343)
	at com.aws.greengrass.shadowmanager.ShadowManager.insertSyncInfoIfNotPresent(ShadowManager.java:437)
	at com.aws.greengrass.shadowmanager.ShadowManager.initializeSyncInfo(ShadowManager.java:432)
	at com.aws.greengrass.shadowmanager.ShadowManager.startSyncingShadows(ShadowManager.java:537)
	at com.aws.greengrass.shadowmanager.ShadowManager.setupSync(ShadowManager.java:396)
	at com.aws.greengrass.shadowmanager.ShadowManager.startup(ShadowManager.java:480)
	at com.aws.greengrass.lifecyclemanager.Lifecycle.lambda$handleStateTransitionStartingToRunningAsync$9(Lifecycle.java:591)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)

2023-06-13T20:35:28.902Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.shadowmanager.ShadowManager: service-set-state. {serviceName=aws.greengrass.ShadowManager, currentState=RUNNING, newState=BROKEN}
2023-06-13T20:35:28.904Z [INFO] (aws.greengrass.ShadowManager-lifecycle) com.aws.greengrass.status.FleetStatusService: fss-status-update-published. Status update published to FSS. {trigger=BROKEN_COMPONENT, serviceName=FleetStatusService, currentState=RUNNING}
```

I am not sure why the `pool size` for the `dbWriteThreadPool` (as defined [here](https://github.com/aws-greengrass/aws-greengrass-shadow-manager/blob/63a00b4d6b83bbae3a05c18ace5213a4a44dcb35/src/main/java/com/aws/greengrass/shadowmanager/ShadowManagerDatabase.java#LL56C18-L56C18)) is logged as 0.

> If you have logs from the issue which didn't involve using local debug console to reinstall, I'd be very interested.

Here are some more logs:

This is what `greengrass.log` contained when my components weren't able to read shadows locally: 
```
2023-06-13T23:23:21.488Z [INFO] (Thread-11) software.amazon.awssdk.eventstreamrpc.RpcServer: New connection code [AWS_ERROR_SUCCESS] for [Id 37, Class ServerConnection, Refs 1](2023-06-13T23:23:21.487736Z) - <null>. {}
2023-06-13T23:23:21.509Z [INFO] (Thread-11) software.amazon.awssdk.eventstreamrpc.ServiceOperationMappingContinuationHandler: aws.greengrass#GreengrassCoreIPC authenticated identity: com.andium.MediaCaptureRoutineManager. {}
2023-06-13T23:23:21.510Z [INFO] (Thread-11) software.amazon.awssdk.eventstreamrpc.ServiceOperationMappingContinuationHandler: Connection accepted for PatricksComponent. {}
2023-06-13T23:23:21.512Z [INFO] (Thread-11) software.amazon.awssdk.eventstreamrpc.ServiceOperationMappingContinuationHandler: Sending connect response for PatricksComponent. {}
2023-06-13T23:23:21.620Z [ERROR] (Thread-11) com.aws.greengrass.ipc.common.ExceptionUtil: Unhandled exception in IPC. {}
java.lang.NullPointerException
        at com.aws.greengrass.shadowmanager.ShadowManagerDAOImpl.getShadowThing(ShadowManagerDAOImpl.java:61)
        at com.aws.greengrass.shadowmanager.ipc.GetThingShadowRequestHandler.lambda$handleRequest$0(GetThingShadowRequestHandler.java:89)
        at com.aws.greengrass.ipc.common.ExceptionUtil.translateExceptions(ExceptionUtil.java:33)
        at com.aws.greengrass.shadowmanager.ipc.GetThingShadowRequestHandler.handleRequest(GetThingShadowRequestHandler.java:75)
        at com.aws.greengrass.shadowmanager.ipc.GetThingShadowIPCHandler.handleRequest(GetThingShadowIPCHandler.java:82)
        at com.aws.greengrass.shadowmanager.ipc.GetThingShadowIPCHandler.handleRequest(GetThingShadowIPCHandler.java:28)
        at software.amazon.awssdk.eventstreamrpc.OperationContinuationHandler.onContinuationMessage(OperationContinuationHandler.java:291)
        at software.amazon.awssdk.crt.eventstream.ServerConnectionContinuationHandler.onContinuationMessageShim(ServerConnectionContinuationHandler.java:53)

2023-06-13T23:23:21.643Z [ERROR] (Thread-11) com.aws.greengrass.ipc.common.ExceptionUtil: Unhandled exception in IPC. {}
java.lang.NullPointerException
        at com.aws.greengrass.shadowmanager.ShadowManagerDAOImpl.getShadowThing(ShadowManagerDAOImpl.java:61)
        at com.aws.greengrass.shadowmanager.ipc.GetThingShadowRequestHandler.lambda$handleRequest$0(GetThingShadowRequestHandler.java:89)
        at com.aws.greengrass.ipc.common.ExceptionUtil.translateExceptions(ExceptionUtil.java:33)
        at com.aws.greengrass.shadowmanager.ipc.GetThingShadowRequestHandler.handleRequest(GetThingShadowRequestHandler.java:75)
        at com.aws.greengrass.shadowmanager.ipc.GetThingShadowIPCHandler.handleRequest(GetThingShadowIPCHandler.java:82)
        at com.aws.greengrass.shadowmanager.ipc.GetThingShadowIPCHandler.handleRequest(GetThingShadowIPCHandler.java:28)
        at software.amazon.awssdk.eventstreamrpc.OperationContinuationHandler.onContinuationMessage(OperationContinuationHandler.java:291)
        at software.amazon.awssdk.crt.eventstream.ServerConnectionContinuationHandler.onContinuationMessageShim(ServerConnectionContinuationHandler.java:53)
```

This is what I saw when restarting, then re-deploying Nucleus, ShadowManager, LocalDebugConsole, and the CLI: 
```
2023-06-14T04:02:20.567Z [INFO] (AwsEventLoop 2) com.aws.greengrass.mqttclient.AwsIotMqtt5Client: Successfully connected to AWS IoT Core. {clientId=3343ffe8-78d0-4d97-ae77-0fd6c1f702b5, sessionPresent=false}
2023-06-14T04:02:20.678Z [INFO] (main) com.aws.greengrass.lifecyclemanager.Kernel: effective-config-dump-complete. {file=/greengrass/v2/config/effectiveConfig.yaml}
2023-06-14T04:02:20.679Z [INFO] (main) com.aws.greengrass.lifecyclemanager.KernelLifecycle: system-start. {main=services.main:NEW}
2023-06-14T04:02:20.689Z [INFO] (DeploymentService-lifecycle) com.aws.greengrass.deployment.DeploymentService: service-set-state. {serviceName=DeploymentService, currentState=NEW, newState=INSTALLED}
2023-06-14T04:02:20.690Z [INFO] (aws.greengrass.Nucleus-lifecycle) com.aws.greengrass.lifecyclemanager.GenericExternalService: service-set-state. {serviceName=aws.greengrass.Nucleus, currentState=NEW, newState=INSTALLED}
2023-06-14T04:02:20.691Z [INFO] (aws.greengrass.LocalDebugConsole-lifecycle) com.aws.greengrass.localdebugconsole.SimpleHttpServer: service-set-state. {serviceName=aws.greengrass.LocalDebugConsole, currentState=NEW, newState=INSTALLED}
2023-06-14T04:02:20.692Z [INFO] (TelemetryAgent-lifecycle) com.aws.greengrass.telemetry.TelemetryAgent: service-set-state. {serviceName=TelemetryAgent, currentState=NEW, newState=INSTALLED}
2023-06-14T04:02:20.692Z [INFO] (FleetStatusService-lifecycle) com.aws.greengrass.status.FleetStatusService: service-set-state. {serviceName=FleetStatusService, currentState=NEW, newState=INSTALLED}
2023-06-14T04:02:20.696Z [INFO] (main-lifecycle) com.aws.greengrass.lifecyclemanager.GenericExternalService: service-set-state. {serviceName=main, currentState=NEW, newState=INSTALLED}
2023-06-14T04:02:20.698Z [INFO] (UpdateSystemPolicyService-lifecycle) com.aws.greengrass.lifecyclemanager.UpdateSystemPolicyService: service-set-state. {serviceName=UpdateSystemPolicyService, currentState=NEW, newState=INSTALLED}
2023-06-14T04:02:20.698Z [INFO] (DeploymentService-lifecycle) com.aws.greengrass.deployment.DeploymentService: service-set-state. {serviceName=DeploymentService, currentState=INSTALLED, newState=STARTING}
2023-06-14T04:02:20.719Z [INFO] (TelemetryAgent-lifecycle) com.aws.greengrass.telemetry.TelemetryAgent: service-set-state. {serviceName=TelemetryAgent, currentState=INSTALLED, newState=STARTING}
2023-06-14T04:02:20.729Z [INFO] (aws.greengrass.Nucleus-lifecycle) com.aws.greengrass.lifecyclemanager.GenericExternalService: service-set-state. {serviceName=aws.greengrass.Nucleus, currentState=INSTALLED, newState=STARTING}
2023-06-14T04:02:20.730Z [INFO] (UpdateSystemPolicyService-lifecycle) com.aws.greengrass.lifecyclemanager.UpdateSystemPolicyService: service-set-state. {serviceName=UpdateSystemPolicyService, currentState=INSTALLED, newState=STARTING}
2023-06-14T04:02:20.731Z [INFO] (FleetStatusService-lifecycle) com.aws.greengrass.status.FleetStatusService: service-set-state. {serviceName=FleetStatusService, currentState=INSTALLED, newState=STARTING}
2023-06-14T04:02:20.733Z [INFO] (aws.greengrass.LocalDebugConsole-lifecycle) com.aws.greengrass.localdebugconsole.SimpleHttpServer: service-set-state. {serviceName=aws.greengrass.LocalDebugConsole, currentState=INSTALLED, newState=STARTING}
2023-06-14T04:02:20.739Z [INFO] (DeploymentService-lifecycle) com.aws.greengrass.deployment.DeploymentService: service-set-state. {serviceName=DeploymentService, currentState=STARTING, newState=RUNNING}
2023-06-14T04:02:20.740Z [INFO] (TelemetryAgent-lifecycle) com.aws.greengrass.telemetry.TelemetryAgent: service-set-state. {serviceName=TelemetryAgent, currentState=STARTING, newState=RUNNING}
2023-06-14T04:02:20.743Z [INFO] (pool-2-thread-13) com.aws.greengrass.lifecyclemanager.GenericExternalService: generic-service-finished. Nothing done. {serviceName=aws.greengrass.Nucleus, currentState=STARTING}
2023-06-14T04:02:20.743Z [INFO] (aws.greengrass.Nucleus-lifecycle) com.aws.greengrass.lifecyclemanager.GenericExternalService: service-set-state. {serviceName=aws.greengrass.Nucleus, currentState=STARTING, newState=FINISHED}
2023-06-14T04:02:20.745Z [INFO] (FleetStatusService-lifecycle) com.aws.greengrass.status.FleetStatusService: service-set-state. {serviceName=FleetStatusService, currentState=STARTING, newState=RUNNING}
2023-06-14T04:02:20.748Z [INFO] (UpdateSystemPolicyService-lifecycle) com.aws.greengrass.lifecyclemanager.UpdateSystemPolicyService: service-set-state. {serviceName=UpdateSystemPolicyService, currentState=STARTING, newState=RUNNING}
2023-06-14T04:02:20.752Z [INFO] (aws.greengrass.Nucleus-lifecycle) com.aws.greengrass.lifecyclemanager.GenericExternalService: service-set-state. {serviceName=aws.greengrass.Nucleus, currentState=FINISHED, newState=STOPPING}
2023-06-14T04:02:20.759Z [INFO] (pool-2-thread-20) com.aws.greengrass.lifecyclemanager.GenericExternalService: Shutdown initiated. {serviceName=aws.greengrass.Nucleus, currentState=STOPPING}
2023-06-14T04:02:20.763Z [INFO] (pool-2-thread-20) com.aws.greengrass.lifecyclemanager.GenericExternalService: generic-service-shutdown. {serviceName=aws.greengrass.Nucleus, currentState=STOPPING}
2023-06-14T04:02:20.764Z [INFO] (aws.greengrass.Nucleus-lifecycle) com.aws.greengrass.lifecyclemanager.GenericExternalService: service-set-state. {serviceName=aws.greengrass.Nucleus, currentState=STOPPING, newState=FINISHED}
2023-06-14T04:02:20.804Z [INFO] (main) com.aws.greengrass.status.FleetStatusService: fss-status-update-published. Status update published to FSS. {trigger=NUCLEUS_LAUNCH, serviceName=FleetStatusService, currentState=RUNNING}
2023-06-14T04:02:20.810Z [INFO] (pool-2-thread-15) org.flywaydb.core.internal.license.VersionPrinter: Flyway Community Edition 6.5.5 by Redgate. {}
2023-06-14T04:02:20.858Z [INFO] (AwsEventLoop 2) com.aws.greengrass.deployment.IotJobsHelper: No deployment job found. {ThingName=3343ffe8-78d0-4d97-ae77-0fd6c1f702b5}
2023-06-14T04:02:20.994Z [INFO] (AwsEventLoop 2) com.aws.greengrass.deployment.ShadowDeploymentListener: Deployment result already reported. Ignoring shadow update at startup. {CONFIGURATION_ARN=arn:aws:greengrass:us-east-1:379266526677:configuration:thing/3343ffe8-78d0-4d97-ae77-0fd6c1f702b5:16}
2023-06-14T04:02:21.580Z [INFO] (pool-2-thread-15) org.flywaydb.core.internal.database.DatabaseFactory: Database: jdbc:h2:/greengrass/v2/work/aws.greengrass.ShadowManager/shadow (H2 1.4). {}
2023-06-14T04:02:21.993Z [INFO] (pool-2-thread-15) org.flywaydb.core.internal.command.DbValidate: Successfully validated 1 migration (execution time 00:00.144s). {}
2023-06-14T04:02:22.085Z [INFO] (pool-2-thread-15) org.flywaydb.core.internal.command.DbMigrate: Current version of schema "PUBLIC": 1. {}
2023-06-14T04:02:22.105Z [INFO] (pool-2-thread-15) org.flywaydb.core.internal.command.DbMigrate: Schema "PUBLIC" is up to date. No migration necessary.. {}
2023-06-14T04:02:22.387Z [ERROR] (pool-2-thread-15) com.aws.greengrass.shadowmanager.ShadowManager: service-errored. {serviceName=aws.greengrass.ShadowManager, currentState=NEW}
java.lang.IllegalArgumentException: configuration in services.aws.greengrass.ShadowManager is already a leaf, cannot become a container
        at com.aws.greengrass.config.Topics.createInteriorChild(Topics.java:165)
        at com.aws.greengrass.config.Topics.createInteriorChild(Topics.java:151)
        at com.aws.greengrass.config.Topics.lookupTopics(Topics.java:233)
        at com.aws.greengrass.config.Topics.lookupTopics(Topics.java:219)
        at com.aws.greengrass.shadowmanager.ShadowManager.install(ShadowManager.java:245)
        at com.aws.greengrass.shadowmanager.ShadowManager.install(ShadowManager.java:225)
        at com.aws.greengrass.lifecyclemanager.Lifecycle.lambda$handleCurrentStateNew$5(Lifecycle.java:502)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
```
For this log, ShadowManager was showing up as in the RUNNING state but IPC to it to retrieve shadows failed. Reinstalling via the debug console led to the stack trace that was posted in the OP.

--- a/src/integrationtests/java/com/aws/greengrass/integrationtests/ShadowManagerDAOImplTest.java
+++ b/src/integrationtests/java/com/aws/greengrass/integrationtests/ShadowManagerDAOImplTest.java
@@ -80,7 +80,6 @@ class ShadowManagerDAOImplTest {
         database = new ShadowManagerDatabase(kernel);
         database.install();
         JsonUtil.loadSchema();
-        database.open();
         dao = new ShadowManagerDAOImpl(database);
     }
 

--- a/src/integrationtests/java/com/aws/greengrass/integrationtests/ShadowManagerDatabaseTest.java
+++ b/src/integrationtests/java/com/aws/greengrass/integrationtests/ShadowManagerDatabaseTest.java
@@ -13,8 +13,9 @@ import com.aws.greengrass.shadowmanager.ShadowManagerDatabase;
 import com.aws.greengrass.shadowmanager.exception.ShadowManagerDataException;
 import com.aws.greengrass.shadowmanager.model.ShadowDocument;
 import com.aws.greengrass.testcommons.testutilities.GGExtension;
-import org.flywaydb.core.internal.exception.FlywaySqlException;
 import org.apache.commons.io.FileUtils;
+import org.flywaydb.core.internal.exception.FlywaySqlException;
+import org.h2.jdbcx.JdbcConnectionPool;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
@@ -66,7 +67,6 @@ import static org.hamcrest.Matchers.lessThanOrEqualTo;
 import static org.hamcrest.Matchers.not;
 import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
-import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.fail;
 import static org.junit.jupiter.params.provider.Arguments.arguments;
 
@@ -185,9 +185,9 @@ class ShadowManagerDatabaseTest extends NucleusLaunchUtils {
         assertNotNull(c, "connection should not be null");
         assertThat("connection is not closed", c.isClosed(), is(false));
         c.close();
+        JdbcConnectionPool pool = db.getPool();
         db.close();
-        assertThat("active connections", db.getPool().getActiveConnections(), is(0));
-        assertThrows(IllegalStateException.class, () -> db.getPool().getConnection());
+        assertThat("active connections", pool.getActiveConnections(), is(0));
     }
 
     @Test

--- a/src/integrationtests/java/com/aws/greengrass/integrationtests/ShadowManagerDatabaseTest.java
+++ b/src/integrationtests/java/com/aws/greengrass/integrationtests/ShadowManagerDatabaseTest.java
@@ -80,7 +80,6 @@ class ShadowManagerDatabaseTest extends NucleusLaunchUtils {
         System.setProperty("aws.greengrass.scanSelfClasspath", "true");
         db = new ShadowManagerDatabase(rootDir);
         db.install();
-        db.open();
     }
 
     @AfterEach
@@ -126,7 +125,6 @@ class ShadowManagerDatabaseTest extends NucleusLaunchUtils {
     @Test
     void GIVEN_migrations_WHEN_install_THEN_shadow_manager_database_installs_and_starts_successfully() throws Exception {
         // GIVEN
-        db.open();
         assertNotNull(db.getPool());
 
         // WHEN
@@ -158,7 +156,6 @@ class ShadowManagerDatabaseTest extends NucleusLaunchUtils {
         Path source = Paths.get(getClass().getResource("database/corrupted.mv.db").toURI());
         Files.copy(source, dest, StandardCopyOption.REPLACE_EXISTING);
         // GIVEN
-        db.open();
         assertNotNull(db.getPool());
 
         // WHEN
@@ -196,7 +193,6 @@ class ShadowManagerDatabaseTest extends NucleusLaunchUtils {
     @Test
     void GIVEN_shadow_manager_database_open_WHEN_closed_and_opened_THEN_shadow_manager_database_can_return_connections() throws Exception {
         db.close();
-        db.open();
         Connection c = assertDoesNotThrow(() -> db.getPool().getConnection());
         assertNotNull(c, "connection should not be null");
         assertThat("connection is not closed", c.isClosed(), is(false));

--- a/src/integrationtests/java/com/aws/greengrass/integrationtests/ShadowManagerTest.java
+++ b/src/integrationtests/java/com/aws/greengrass/integrationtests/ShadowManagerTest.java
@@ -42,7 +42,6 @@ import org.junit.jupiter.api.extension.ExtensionContext;
 import org.mockito.junit.jupiter.MockitoExtension;
 import software.amazon.awssdk.services.iotdataplane.model.GetThingShadowResponse;
 
-import javax.net.ssl.KeyManager;
 import java.time.Instant;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -53,6 +52,7 @@ import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
+import javax.net.ssl.KeyManager;
 
 import static com.aws.greengrass.componentmanager.KernelConfigResolver.CONFIGURATION_CONFIG_KEY;
 import static com.aws.greengrass.deployment.DeviceConfiguration.DEVICE_PARAM_THING_NAME;
@@ -66,7 +66,6 @@ import static com.aws.greengrass.shadowmanager.model.Constants.CONFIGURATION_STR
 import static com.aws.greengrass.shadowmanager.model.Constants.CONFIGURATION_SYNCHRONIZATION_TOPIC;
 import static com.aws.greengrass.shadowmanager.model.Constants.CONFIGURATION_SYNC_DIRECTION_TOPIC;
 import static com.aws.greengrass.shadowmanager.model.Constants.CONFIGURATION_THING_NAME_TOPIC;
-
 import static com.aws.greengrass.testcommons.testutilities.ExceptionLogProtector.ignoreExceptionOfType;
 import static com.github.grantwest.eventually.EventuallyLambdaMatcher.eventuallyEval;
 import static org.hamcrest.MatcherAssert.assertThat;

--- a/src/integrationtests/java/com/aws/greengrass/integrationtests/ShadowManagerTest.java
+++ b/src/integrationtests/java/com/aws/greengrass/integrationtests/ShadowManagerTest.java
@@ -169,6 +169,9 @@ class ShadowManagerTest extends NucleusLaunchUtils {
                 .configFile(DEFAULT_CONFIG)
                 .mqttConnected(false)
                 .build());
+        // Validate that shadow manager will continue to work properly even if it is reinstalled.
+        shadowManager.requestReinstall();
+        Thread.sleep(1000);
         shadowManager.startSyncingShadows(ShadowManager.StartSyncInfo.builder().build());
         ShadowManagerDAOImpl impl = kernel.getContext().get(ShadowManagerDAOImpl.class);
         createThingShadowSyncInfo(impl, THING_NAME);

--- a/src/main/java/com/aws/greengrass/shadowmanager/ShadowManager.java
+++ b/src/main/java/com/aws/greengrass/shadowmanager/ShadowManager.java
@@ -474,8 +474,6 @@ public class ShadowManager extends PluginService {
     @SuppressWarnings({"PMD.AvoidCatchingGenericException"})
     public void startup() {
         try {
-            database.open();
-
             reportState(State.RUNNING);
             setupSync();
         } catch (Exception e) {
@@ -484,13 +482,14 @@ public class ShadowManager extends PluginService {
     }
 
     @Override
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
     protected void shutdown() throws InterruptedException {
         try {
             stopSyncingShadows(true);
             pubSubIntegrator.unsubscribe();
             inboundRateLimiter.clear();
             database.close();
-        } catch (IOException e) {
+        } catch (Exception e) {
             logger.atError()
                     .setEventType(LogEvents.DATABASE_CLOSE_ERROR.code())
                     .setCause(e)

--- a/src/main/java/com/aws/greengrass/shadowmanager/ShadowManagerDAOImpl.java
+++ b/src/main/java/com/aws/greengrass/shadowmanager/ShadowManagerDAOImpl.java
@@ -12,6 +12,7 @@ import com.aws.greengrass.shadowmanager.exception.ShadowManagerDataException;
 import com.aws.greengrass.shadowmanager.model.ShadowDocument;
 import com.aws.greengrass.shadowmanager.model.dao.SyncInformation;
 import com.aws.greengrass.util.Pair;
+import org.h2.jdbcx.JdbcConnectionPool;
 
 import java.io.IOException;
 import java.sql.Connection;
@@ -58,7 +59,7 @@ public class ShadowManagerDAOImpl implements ShadowManagerDAO {
         String sql = "SELECT document, version, updateTime FROM documents  WHERE deleted = 0 AND "
                 + "thingName = ? AND shadowName = ?";
 
-        try (Connection c = database.getPool().getConnection();
+        try (Connection c = getPool().getConnection();
              PreparedStatement preparedStatement = c.prepareStatement(sql)) {
             preparedStatement.setString(1, thingName);
             preparedStatement.setString(2, shadowName);
@@ -73,6 +74,15 @@ public class ShadowManagerDAOImpl implements ShadowManagerDAO {
         } catch (SQLException | IOException | IllegalStateException e) {
             throw new ShadowManagerDataException(e);
         }
+    }
+
+    private JdbcConnectionPool getPool() {
+        JdbcConnectionPool pool = database.getPool();
+        if (pool == null) {
+            throw new ShadowManagerDataException("Database pool not initialized. Shadow manager most likely isn't "
+                    + "running yet. Wait for Shadow manager to be running.");
+        }
+        return pool;
     }
 
     /**
@@ -262,7 +272,7 @@ public class ShadowManagerDAOImpl implements ShadowManagerDAO {
     public Optional<Long> getDeletedShadowVersion(String thingName, String shadowName) {
         String sql = "SELECT version FROM documents  WHERE deleted = 1 AND thingName = ? AND shadowName = ?";
 
-        try (Connection c = database.getPool().getConnection();
+        try (Connection c = getPool().getConnection();
              PreparedStatement preparedStatement = c.prepareStatement(sql)) {
             preparedStatement.setString(1, thingName);
             preparedStatement.setString(2, shadowName);
@@ -359,7 +369,7 @@ public class ShadowManagerDAOImpl implements ShadowManagerDAO {
     }
 
     private <T> T execute(String sql, SQLExecution<T> thunk) {
-        try (Connection c = database.getPool().getConnection();
+        try (Connection c = getPool().getConnection();
              PreparedStatement statement = c.prepareStatement(sql)) {
             statement.setQueryTimeout(10);
             return thunk.apply(statement);

--- a/src/main/java/com/aws/greengrass/shadowmanager/ShadowManagerDAOImpl.java
+++ b/src/main/java/com/aws/greengrass/shadowmanager/ShadowManagerDAOImpl.java
@@ -77,12 +77,11 @@ public class ShadowManagerDAOImpl implements ShadowManagerDAO {
     }
 
     private JdbcConnectionPool getPool() {
-        JdbcConnectionPool pool = database.getPool();
-        if (pool == null) {
+        if (!database.isInitialized()) {
             throw new ShadowManagerDataException("Database pool not initialized. Shadow manager most likely isn't "
                     + "running yet. Wait for Shadow manager to be running.");
         }
-        return pool;
+        return database.getPool();
     }
 
     /**

--- a/src/main/java/com/aws/greengrass/shadowmanager/ShadowManagerDatabase.java
+++ b/src/main/java/com/aws/greengrass/shadowmanager/ShadowManagerDatabase.java
@@ -53,7 +53,7 @@ public class ShadowManagerDatabase implements Closeable {
     private static final Logger logger = LogManager.getLogger(ShadowManagerDatabase.class);
     private final Path databasePath;
     @Getter
-    private final ExecutorService dbWriteThreadPool = Executors.newCachedThreadPool();
+    private ExecutorService dbWriteThreadPool = Executors.newCachedThreadPool();
 
     /**
      * Creates a database with a {@link javax.sql.DataSource} using the kernel config.
@@ -136,6 +136,11 @@ public class ShadowManagerDatabase implements Closeable {
      */
     @Synchronized
     public void open() {
+        // The DB can be closed and then opened again, so we need to create a new threadpool if this happens
+        // because we shut it down in the close().
+        if (dbWriteThreadPool == null || dbWriteThreadPool.isShutdown()) {
+            dbWriteThreadPool = Executors.newCachedThreadPool();
+        }
         if (closed) {
             // defaults to 10 connections with a 30s timeout waiting for a connection
             pool = JdbcConnectionPool.create(dataSource);

--- a/src/main/java/com/aws/greengrass/shadowmanager/ShadowManagerDatabase.java
+++ b/src/main/java/com/aws/greengrass/shadowmanager/ShadowManagerDatabase.java
@@ -10,6 +10,7 @@ import com.aws.greengrass.logging.api.Logger;
 import com.aws.greengrass.logging.impl.LogManager;
 import com.aws.greengrass.shadowmanager.exception.ShadowManagerDataException;
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import lombok.Getter;
 import lombok.Synchronized;
 import org.flywaydb.core.Flyway;
 import org.flywaydb.core.api.FlywayException;
@@ -41,7 +42,7 @@ public class ShadowManagerDatabase implements Closeable {
     // these setting optimize for minimal disk space over concurrent performance
     private static final String DATABASE_FORMAT = "jdbc:h2:%s/%s"
             + ";RETENTION_TIME=1000" // ms - time to keep values for before writing to disk (default is 45000)
-            + ";DEFRAG_ALWAYS=FALSE" // don't defragment db on shutdown because that's a risky time to change the db
+            + ";DEFRAG_ALWAYS=TRUE" // defragment db on shutdown (ensures only a single value in db on close)
             + ";COMPRESS=TRUE" // compress large objects (clob/blob) (default false)
             ;
     private final JdbcDataSource dataSource;
@@ -51,6 +52,8 @@ public class ShadowManagerDatabase implements Closeable {
     private static final Logger logger = LogManager.getLogger(ShadowManagerDatabase.class);
     private final Path databasePath;
     private ExecutorService dbWriteThreadPool;
+    @Getter
+    private boolean initialized = false;
 
     /**
      * Creates a database with a {@link javax.sql.DataSource} using the kernel config.
@@ -87,6 +90,7 @@ public class ShadowManagerDatabase implements Closeable {
                 deleteDB(databasePath);
                 migrateDB();
             }
+            initialized = true;
         } catch (FlywayException | IOException e) {
             throw new ShadowManagerDataException(e);
         }

--- a/src/main/java/com/aws/greengrass/shadowmanager/ShadowManagerDatabase.java
+++ b/src/main/java/com/aws/greengrass/shadowmanager/ShadowManagerDatabase.java
@@ -10,7 +10,6 @@ import com.aws.greengrass.logging.api.Logger;
 import com.aws.greengrass.logging.impl.LogManager;
 import com.aws.greengrass.shadowmanager.exception.ShadowManagerDataException;
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
-import lombok.Getter;
 import lombok.Synchronized;
 import org.flywaydb.core.Flyway;
 import org.flywaydb.core.api.FlywayException;
@@ -23,6 +22,7 @@ import java.io.Closeable;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.sql.Connection;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.stream.Stream;
@@ -41,19 +41,16 @@ public class ShadowManagerDatabase implements Closeable {
     // these setting optimize for minimal disk space over concurrent performance
     private static final String DATABASE_FORMAT = "jdbc:h2:%s/%s"
             + ";RETENTION_TIME=1000" // ms - time to keep values for before writing to disk (default is 45000)
-            + ";DEFRAG_ALWAYS=TRUE" // defragment db on shutdown (ensures only a single value in db on close)
+            + ";DEFRAG_ALWAYS=FALSE" // don't defragment db on shutdown because that's a risky time to change the db
             + ";COMPRESS=TRUE" // compress large objects (clob/blob) (default false)
             ;
     private final JdbcDataSource dataSource;
 
-    @Getter
     private JdbcConnectionPool pool;
 
-    private boolean closed = true;
     private static final Logger logger = LogManager.getLogger(ShadowManagerDatabase.class);
     private final Path databasePath;
-    @Getter
-    private ExecutorService dbWriteThreadPool = Executors.newCachedThreadPool();
+    private ExecutorService dbWriteThreadPool;
 
     /**
      * Creates a database with a {@link javax.sql.DataSource} using the kernel config.
@@ -103,18 +100,40 @@ public class ShadowManagerDatabase implements Closeable {
         flyway.migrate();
     }
 
+    @SuppressWarnings({"checkstyle:EmptyBlock", "checkstyle:WhitespaceAround", "PMD.AvoidCatchingGenericException"})
     private boolean migrateAndGetResult() {
         try {
             migrateDB();
-            return true;
         } catch (FlywaySqlException flywaySqlException) {
             if (flywaySqlException.getCause() instanceof JdbcSQLNonTransientException
                     && flywaySqlException.getCause().getCause() instanceof IllegalStateException) {
-                logger.atError().cause(flywaySqlException).log("Shadow manager DB is corrupted.");
+                logger.atError().cause(flywaySqlException).log("Shadow manager DB is corrupted");
                 return false;
             }
             throw flywaySqlException;
         }
+
+        // Validate that after migration we're actually able to open and connect to the DB.
+        // This may fail if closing the DB after migration failed for some reason.
+        try {
+            try (Connection p = getPool().getConnection()) {}
+            return true;
+        } catch (Exception e) {
+            logger.atError().cause(e).log("Shadow manager DB could not be opened. Deleting and recreating it");
+            close();
+            return false;
+        }
+    }
+
+    /**
+     * Get a reference to the connection pool.
+     * @return JDBC connection pool
+     */
+    public synchronized JdbcConnectionPool getPool() {
+        if (pool == null) {
+            pool = JdbcConnectionPool.create(dataSource);
+        }
+        return pool;
     }
 
     private void deleteDB(Path databasePath) throws IOException {
@@ -131,31 +150,29 @@ public class ShadowManagerDatabase implements Closeable {
         }
     }
 
-    /**
-     * Open the database to allow connections.
-     */
+    @Override
     @Synchronized
-    public void open() {
-        // The DB can be closed and then opened again, so we need to create a new threadpool if this happens
-        // because we shut it down in the close().
-        if (dbWriteThreadPool == null || dbWriteThreadPool.isShutdown()) {
-            dbWriteThreadPool = Executors.newCachedThreadPool();
+    @SuppressWarnings("PMD.NullAssignment")
+    @SuppressFBWarnings(value = "UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR", justification = "Field gated by flag")
+    public void close() {
+        if (dbWriteThreadPool != null) {
+            dbWriteThreadPool.shutdown();
         }
-        if (closed) {
-            // defaults to 10 connections with a 30s timeout waiting for a connection
-            pool = JdbcConnectionPool.create(dataSource);
-            closed = false;
+        if (pool != null) {
+            pool.dispose();
+            pool = null;
         }
     }
 
-    @Override
-    @Synchronized
-    @SuppressFBWarnings(value = "UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR", justification = "Field gated by flag")
-    public void close() throws IOException {
-        if (!closed) {
-            dbWriteThreadPool.shutdown();
-            pool.dispose();
-            closed = true;
+    /**
+     * Get the thread pool used for writing to the DB.
+     *
+     * @return a running thread pool
+     */
+    public synchronized ExecutorService getDbWriteThreadPool() {
+        if (dbWriteThreadPool == null || dbWriteThreadPool.isShutdown()) {
+            dbWriteThreadPool = Executors.newCachedThreadPool();
         }
+        return dbWriteThreadPool;
     }
 }

--- a/src/main/java/com/aws/greengrass/shadowmanager/exception/ShadowManagerDataException.java
+++ b/src/main/java/com/aws/greengrass/shadowmanager/exception/ShadowManagerDataException.java
@@ -11,4 +11,8 @@ public class ShadowManagerDataException extends RuntimeException {
     public ShadowManagerDataException(final Throwable ex) {
         super(ex);
     }
+
+    public ShadowManagerDataException(final String s) {
+        super(s);
+    }
 }

--- a/src/test/java/com/aws/greengrass/shadowmanager/ShadowManagerDAOImplTest.java
+++ b/src/test/java/com/aws/greengrass/shadowmanager/ShadowManagerDAOImplTest.java
@@ -93,6 +93,7 @@ class ShadowManagerDAOImplTest {
         lenient().when(mockDatabase.getDbWriteThreadPool()).thenReturn(es);
         when(mockDatabase.getPool()).thenReturn(mockPool);
         when(mockPool.getConnection()).thenReturn(mockConnection);
+        when(mockDatabase.isInitialized()).thenReturn(true);
         JsonUtil.loadSchema();
     }
     @AfterEach

--- a/src/test/java/com/aws/greengrass/shadowmanager/ShadowManagerUnitTest.java
+++ b/src/test/java/com/aws/greengrass/shadowmanager/ShadowManagerUnitTest.java
@@ -617,7 +617,6 @@ class ShadowManagerUnitTest extends GGServiceTestUtil {
         verify(mockCloudDataClient, times(1)).updateSubscriptions(anySet());
         verify(mockSyncHandler, times(1)).start(any(SyncContext.class), anyInt());
 
-        verify(mockDatabase, times(1)).open();
         verify(mockDao, times(1)).deleteSyncInformation("foo", "bar");
 
         ArgumentCaptor<SyncInformation> captor = ArgumentCaptor.forClass(SyncInformation.class);

--- a/src/test/java/com/aws/greengrass/shadowmanager/ShadowManagerUnitTest.java
+++ b/src/test/java/com/aws/greengrass/shadowmanager/ShadowManagerUnitTest.java
@@ -97,7 +97,6 @@ import static org.mockito.Mockito.atMostOnce;
 import static org.mockito.Mockito.doAnswer;
 import static org.mockito.Mockito.doNothing;
 import static org.mockito.Mockito.doReturn;
-import static org.mockito.Mockito.doThrow;
 import static org.mockito.Mockito.lenient;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.never;
@@ -645,13 +644,6 @@ class ShadowManagerUnitTest extends GGServiceTestUtil {
         verify(mockDatabase, atMostOnce()).close();
         verify(mockInboundRateLimiter, atMostOnce()).clear();
         verify(mockPubSubClientWrapper, atMostOnce()).unsubscribe(any());
-    }
-
-    @Test
-    void GIVEN_shadow_manager_db_WHEN_shutdown_throws_io_exception_THEN_catches_exception(ExtensionContext extensionContext) throws IOException {
-        ignoreExceptionOfType(extensionContext, IOException.class);
-        doThrow(IOException.class).when(mockDatabase).close();
-        assertDoesNotThrow(() -> shadowManager.shutdown());
     }
 
     @Test

--- a/uat/testing-features/pom.xml
+++ b/uat/testing-features/pom.xml
@@ -41,7 +41,7 @@
         <dependency>
             <groupId>com.aws.greengrass</groupId>
             <artifactId>aws-greengrass-testing-standalone</artifactId>
-            <version>1.0.0-SNAPSHOT</version>
+            <version>1.2.0-SNAPSHOT</version>
             <scope>compile</scope>
         </dependency>
         <dependency>

